docs/proposals/20221611-workload-identity-integration.md

```diff
@@ -0,0 +1,424 @@
+
+			// setCredentialsForWorkloadIdentity just setups the 
+			// PublicCloud env URLs 
+			params.AzureClients.setCredentialsForWorkloadIdentity(ctx, params.AzureCluster.Spec.SubscriptionID, params.AzureCluster.Spec.AzureEnvironment)
+			client.Authorizer = azidext.NewTokenCredentialAdapter(cred, []string{"https://management.azure.com//.default"})
+			params.AzureClients.Authorizer = client.Authorizer
+
+```
+
+**NOTE:**
+`azidext.NewTokenCredentialAdapter` is used to get a authoriser in order to add to the existing code workflow to adapt an azcore.TokenCredential type to an autorest.Authorizer type.
+
+Also a go file e.g `workload_identity.go` in the `identity` package dealing with AZWI functionality.
+
+```go
+
+type workloadIdentityCredential struct {
+	assertion string
+	file      string
+	cred      *azidentity.ClientAssertionCredential
+	lastRead  time.Time
+}
+
+type workloadIdentityCredentialOptions struct {
+	azcore.ClientOptions
+}
+
+func newWorkloadIdentityCredential(tenantID, clientID, file string, options *workloadIdentityCredentialOptions) (*workloadIdentityCredential, error) {
+	w := &workloadIdentityCredential{file: file}
+	cred, err := azidentity.NewClientAssertionCredential(tenantID, clientID, w.getAssertion, &azidentity.ClientAssertionCredentialOptions{ClientOptions: options.ClientOptions})
+	if err != nil {
+		return nil, err
+	}
+	w.cred = cred
+	return w, nil
+}
+
+func (w *workloadIdentityCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (azcore.AccessToken, error) {
+	return w.cred.GetToken(ctx, opts)
+}
+
+func (w *workloadIdentityCredential) getAssertion(context.Context) (string, error) {
+	if now := time.Now(); w.lastRead.Add(5 * time.Minute).Before(now) {
+		content, err := os.ReadFile(w.file)
+		if err != nil {
+			return "", err
+		}
+		w.assertion = string(content)
+		w.lastRead = now
+	}
+	return w.assertion, nil
+}
+
+```
+
+### <a name='OpenQuestions'></a>Open Questions
+
+#### <a name='Howtomultitenancy'></a>1. How to achieve multi-tenancy?
+
+The identity is tied to the client ID which can be supplied via AzureClusterIdentity. In case identity is not found in AzureClusterIdentity client ID from the service account annotation
+`azure.workload.identity/client-id` will be used.
+
+#### <a name='Howtodistributekeys'></a>2. How to distribute key pair to management cluster?
+
+Keys can be distributed by using the CABPK feature by creating a secret with name `<cluster-name>-sa` encompassing the key pair. More details on it [here](https://cluster-api.sigs.k8s.io/tasks/certs/using-custom-certificates.html)
+
+#### <a name='UserExperience'></a>3. User Experience
+
+Though AZWI has a lot of advantages as compared to AAD pod identity, setting up AZWI involves couple of manual step and it can impact the user experience.
+
+### <a name='Limitations'></a>Limitations
+
+- Cloud provider azure does not support workload identity yet and it also needs to authenticate to azure APIs. 
```

Comment:
https://github.com/kubernetes-sigs/cloud-provider-azure/commit/8d912fad87443dba2e2c3950a4d283fb4d718d31 merged but is not yet part of a release

Do we have a good idea of how it will be leveraged in CAPZ once it's supported? Will we need to install workload identity on the workload cluster before cloud-provider? if so how would that work since cloud-provider needs to be installed before the nodes can be "Ready" and Schedulable? 

@sonasingh46 @aramase

Reply:
If I am not wrong, only change that will be required is deploying cloud provider azure with workload identity label and rest will be taken care of. 
I will let @aramase also comment.

Reply:
>if so how would that work since cloud-provider needs to be installed before the nodes can be "Ready" and Schedulable?

Hmm.. This is a good point. Let me try experimenting with this. I can build a cloud provider image that has the workload identity support.

Reply:
Does that warrant workload identity to become a Daemonset? This is what I can think off while typing.

Reply:
Also when I think more about this. The defined approach for forming a management cluster is below. 
BootstrapCluster(On Kind, cloud provider not needed) --> ManagementCluster(cloud provider needed) 

As far as workload identity is considered, the defined/recommended approach will be : 
- Create a bootstrap cluster on kind using manual service principal.
- Use capi pivoting to create a management cluster on Azure. This still uses manual service principal. 
- Someone from the sec ops team migrates the management cluster to be using workload identity and the steps will be mentioned in the capz doc.

Above steps will not have the problem stated above.

Reply:
So if someone is using an existing cluster, the nodes will be ready, no? And then they can just perform the migration steps.
( Am I missing anything? )

Reply:
> Do we have a good idea of how it will be leveraged in CAPZ once it's supported? Will we need to install workload identity on the workload cluster before cloud-provider? if so how would that work since cloud-provider needs to be installed before the nodes can be "Ready" and Schedulable?

This is a chicken and egg situation. The WI webhook is **not required** to use WIF. In this case, the cloud provider could just add the projected service account token volume as part of their pod spec by default and rely on their configuration to get the tenant id, client id and aad endpoint to use.